### ISSUE TITLE
"Populate" extension method return type change proposition

### ISSUE THREAD

**Issue by rafalkwol** [autofac/Autofac.Extensions.DependencyInjection#37]
I have a proposition for a change in the extension methods provided in the package. The issue I encountered is that the Populate method located in the AutofacRegistration class under the covers registers all the services available in the IServiceCollection provided as the parameter. It's a good approach if you use the package solely in ASP.NET Core environment where you usually want the default registrations configurations.

The problem begins when you want to use IServiceCollection in another environment - let's say Windows Service environment. I use it to have the benefit of extensions method provided for IServiceCollection (AddHttpClient<T> in this case). The problem is my class implements IDisposable which makes Autofac automatically assume I want the container to deal with the service disposal. When I'm registering something manually I can use ExternallyOwned extension method, but as the registrations are handled by the package I have no realistic way to override the registration in any non-hacky way. It's only a simple example but the same issue arise if I want to change the scope of the services.

So my proposition would be to simply give the Populate method a return type which would be a collection of IRegistrationBuilder<T, T1, T2> or just objects. In this collection would be all registrations added by the Populate method. This way we would be able to access the registrations and modify them.

Is this kind of change possible?

**Comment by tillig** [comment]
TLDR: I don't think this is going to happen, and I'll explain why.

First, you should know that having the container/scope track and dispose of `IDisposable` objects [is actually part of the specification for `Microsoft.Extensions.DependencyInjection`](https://github.com/aspnet/Extensions/blob/8992b1fa839e8e06e4a556fd6591bb57eb865ebe/src/DependencyInjection/DI.Specification.Tests/src/DependencyInjectionSpecificationTests.cs#L406).

Basically, by using a conforming container model like this, they add some benefits ("standard registrations" like `AddHttpClient` and some others) but they also remove some benefits (like having the tight control you want in some areas or being flexible with things).

Which is to say, if you use the conforming container helpers, the assumption is (and should be) that you're also accepting the drawbacks to using them.

However, even if we did return a set of registrations, the `<T, T1, T2>` would all be different based on whether it's a lambda, provided value, or type mapping. The collection returned would have to be so generic as to be nearly unusable since you'd have to reverse-engineer casting those types to something that would be usable... then you'd also have to query them to find the one you want, then you'd have to make sure it's the right one in the right order (order matters!)... it wouldn't really be that helpful.

What I would recommend instead if you need Autofac-specific features is to not use the conforming container helpers and register things directly with Autofac. Unfortunately, that's not as straightforward and may require a little reverse-engineering on your part to see exactly what the helpers are doing to ensure you're replicating everything you need.

Or, in this particular case, consider resolving `Owned<T>` which is the client half of `ExternallyOwned`.

**Comment by rafalkwol** [comment]
Thanks for explanation. I also assumed a collection of generic registrations may not be an easy thing to provide - hence the proposition of objects collection. Nevertheless I understand what you mean and I guess I will just have to adjust my objects a bit.

For now I created proxy objects which are registered as HttpClients which dont have IDisposable and than I inject those to the final class and forward HttpClient property.

When it comes to Owned<T> I used it with combination to DbContext from Entity Framework Core as it had a scope lifetime. I planned to use it also for HttpClient, but didn't want to abuse it. Any experience with how much of an overhead is to use Owned<T> in medium to large applications?
